grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -352,4 +360,75 @@ class PrepaymentInvoiceService {
 
         return quantityInvoiced >= quantityInverseable ? quantityInverseable : quantityInvoiced
     }
+
+    /**
+     * Updates invoice item's unit price and amount and finds related inverse item and updates unit price and amount
+     * */
+    void updateInvoiceItemUnitPrice(String itemId, BigDecimal unitPrice) {
+        def g = grailsApplication.mainContext.getBean('org.grails.plugins.web.taglib.ApplicationTagLib')
```

Comment:
you can also move the to the separate method, like it is already done in, for example:
![image](https://github.com/user-attachments/assets/30fbb8a6-1507-4e6d-a1e6-f2e567528219)

+ I think using class name instead of strings should be better

Reply:
Actually considering @EWaterman comment in the other pr, I am wondering if I should not just return the code, and the front-end should just handle the translation of it, but I don't remember if this is handled properly right there or not.

Reply:
moving it to a method would at least make it easier to mock in a unit test and is likely good enough for this PR.

I do think we should have a larger group discussion around how localization should work in the backend. My thoughts are that the frontend should handle fetching the localized string and the backend can just return the code, args, and default (minus the "message" tag lib). Since we're planning on eventually splitting the frontend and backend, it seems strange to be working with tag libs at all from the backend. I'd rather have the backend be simple and just return the errors to the client, and then the client can display those errors in whatever way it wants (ie localize them, render them nicely...) but that can be a tech huddle discussion.

Reply:
Actually, I checked the validate exception handling (ErrorsController) and we are relying on ApplicaitonTagLib there as well, so I guess it can stay as is for now. I just moved it out to the method as suggested by Alan.

Reply:
Yeah the reason it's happening there is because it's a backend-driven redirect built into Grails. I haven't thought  through it enough to have a super strong opinion but we might eventually need to think through a different approach if the goal is to have the backend totally split from the frontend since we won't want backend APIs rendering views.

But yeah, that's fine for now.